src/wp-includes/blocks.php

```diff
@@ -1002,6 +1002,115 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+ * @return string The serialized markup with the ignored hooked blocks metadata applied.
+ */
+function apply_hooked_blocks_to_content( $content, $context, $callback = 'insert_hooked_blocks' ) {
+	$hooked_blocks = get_hooked_blocks();
+	if ( empty( $hooked_blocks ) && ! has_filter( 'hooked_block_types' ) ) {
+		return $content;
+	}
+
+	$blocks = parse_blocks( $content );
+
+	$before_block_visitor = make_before_block_visitor( $hooked_blocks, $context, $callback );
+	$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $context, $callback );
+
+	return traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+}
+
+/**
+ * Accepts the serialized markup of a block and its inner blocks, and returns serialized markup of the inner blocks.
+ *
+ * @since 6.6.0
+ * @access private
+ *
+ * @param string $serialized_block The serialized markup of a block and its inner blocks.
+ * @return string
+ */
+function remove_serialized_parent_block( $serialized_block ) {
+	$start = strpos( $serialized_block, '-->' ) + strlen( '-->' );
+	$end   = strrpos( $serialized_block, '<!--' );
+	return substr( $serialized_block, $start, $end - $start );
+}
+
+/**
+ * Updates the wp_postmeta with the list of ignored hooked blocks where the inner blocks are stored as post content.
+ * Currently only supports `wp_navigation` post types.
+ *
+ * @since 6.6.0
+ * @access private
+ *
+ * @param stdClass $post Post object.
+ * @return stdClass The updated post object.
+ */
+function update_ignored_hooked_blocks_postmeta( $post ) {
+	/*
+	 * In this scenario the user has likely tried to create a navigation via the REST API.
+	 * In which case we won't have a post ID to work with and store meta against.
+	 */
+	if ( empty( $post->ID ) ) {
+		return $post;
+	}
+
+	/**
+	 * Skip meta generation when consumers intentionally update specific Navigation fields and omit the content update.
+	 * Or if the incoming post is not a Navigation post type.
+	 */
+	if ( ! isset( $post->post_content ) || ( isset( $post->post_type ) && 'wp_navigation' !== $post->post_type ) ) {
+		return $post;
+	}
+
+	$blocks     = parse_blocks( $post->post_content );
+	$attributes = array();
+
+	if ( isset( $post->ID ) ) {
+		$ignored_hooked_blocks = get_post_meta( $post->ID, '_wp_ignored_hooked_blocks', true );
+		if ( ! empty( $ignored_hooked_blocks ) ) {
+			$ignored_hooked_blocks  = json_decode( $ignored_hooked_blocks, true );
+			$attributes['metadata'] = array(
+				'ignoredHookedBlocks' => $ignored_hooked_blocks,
+			);
+		}
+	}
+
+	$mocked_parent_block = array(
+		'blockName'    => 'core/navigation',
+		'attrs'        => $attributes,
+		'innerBlocks'  => $blocks,
+		'innerContent' => array_fill( 0, count( $blocks ), null ),
+	);
+
+	$serialized_block = apply_hooked_blocks_to_content( traverse_and_serialize_block( $mocked_parent_block ), get_post( $post->ID ), 'set_ignored_hooked_blocks_metadata' );
+	$root_block       = parse_blocks( $serialized_block )[0];
```

Comment:
We'll need to optimize this a bit 😬 Along with [this `parse_blocks` call](https://github.com/WordPress/wordpress-develop/pull/6604/files#diff-8c99af92e4ec0fdb307ddd9b42be1e1ef1efe4a9f31287c23f346244dddd1ce9R1074), we're:

1. Parsing `post_content` (to attach the resulting blocks to the mocked parent block),
2. traversing and reserializing it,
3. calling `apply_hooked_blocks_to_content`, which
    1. parses the serialized markup, and
    2. traverses and reserializes it
4. parsing the resulting markup (to get the root block).

Since parsing and traversal/serialization are expensive operations, let's try to cancel out as many of them as possible. Roughly speaking, I think we can construct the mock parent block without needing to parse `post_content` at that stage, but instead do it on a markup base that we then feed to `apply_hooked_blocks_to_content`.

I think the `get_comment_delimited_block_content()` function should help with that. Something like

```php
$markup = get_comment_delimited_block_content(
	'core/navigation',
	$attributes,
	$post->post_content
);
$serialized_block = apply_hooked_blocks_to_content( $markup, get_post( $post->ID ), 'set_ignored_hooked_blocks_metadata' );
```

(untested)

Reply:
Nice, this does work as expected https://github.com/WordPress/wordpress-develop/pull/6604/commits/53033c4d015b869df9865b3bfc602782646eefd1

I'd love to optimize this further if I can find opportunity to do so. I'm going to focus on some tests to support it before doing so though.